platform/engineering/design-docs/2020-08-11-redirects-from-cms.md

```diff
@@ -0,0 +1,224 @@
+
+### Background
+If pages are moved within the CMS under the right conditions the end result is a combination of internal broken links that currently cause failure of CMS content to be released to the front-end, and 404 errors for veterans.
+
+Within the CMS we have put in a lot of [safeguards to detect broken links](https://www.drupal.org/project/node_link_report) within content at the instant of saving a page of content and warning the edtitor in real time.  There are limits to this method however as it can not account for editors changing the location of a page being linked to or the published status of the link target at the time of content release (CR).
+
+| Situation | Current Result | Result if Redirects as Content |
+| --------- | -------------- | ------------------------------ |
+| Link correctly created in CMS to page in CMS that has been moved | 200 | 200 |
+| Link created by url copy paste in WYSIWG to page in CMS that has been moved | 404 & CR Fail | 200|
+| Link from VA.gov page not in CMS to CMS page that has been moved | 404 & CR Fail | 200 |
+| Link from external site to page in CMS that has been moved | 404 | 200 |
+| Browser bookmark / email link to page in CMS that has been moved | 404 | 200 |
+
+The current state of managing redirects for content, involves content editors and a developer coordinating and choreographing the release of the contnet with the addition of the redirect(s).  It is time intensive and risk prone.  VA needs a redirect system that is easier to manage and less reliant on developers timing releases perfectly to match.
+
+### Guiding Principles
+1. Veterans should not get 404 errors.
+2. Content on VA.gov should not have broken links to content also on VA.gov.
+3. As the CMS scales from ~20 editors and daily content releases, up to ~500 editors and continuous content releases, we can not allow a broken link to prevent content releases.
+4. Developers currently waste far too much time chasing down broken links.
+
+
+### High Level Design
+
+Three parts of the system work together:
+
+#### I. CMS
+At the highest level, the CMS manages redirects already.  It tracks them, sorts out and prevents circular references (infinite redirects), creates new redirects for moved pages and can handle query parameters, if needed for legacy handling. The CMS even provides a nice user interface for adding, deleting, and editing source, destination and status codes.  The CMS currently makes redirects available to both GraphQL and the static file content exports.  Only one minor change is needed on the CMS side to translate node/# paths to the proper alias.  We simply need a way for the Front-End to act on the redirects in the CMS. The CMS treats redirects like content, and the Front-End should too.  When a title changes, in most cases, the URL changes, so a redirect is needed.  Any Front-End system that does not account for that is a 404 factory, which is our current condition.
```

Comment:
> The CMS treats redirects like content, and the Front-End should too.

what does it mean when you say the FE should "treat redirects like content"?

Reply:
I mean that when content gets updated, redirects should get updated.  When the title of a page in the cms changes, its path changes.  So the redirects that Drupal already builds in the CMS, should get built on the front end too at the time of content release.

Reply:
We should be able to decouple URL changes from H1 changes. H1s may evolve and be tweaked but URLs are not directly tied to them for core benefit content.  This approach may be different for other types of content (i.e. news articles) but that needs to be defined yet.

Reply:
Does the CMS eliminate multiple jumps or is that something that needs to be manually managed?  In other words, if a URL A changes to B and A is redirected, then B changes to C, we want the original A URL to redirect to C, not B and then C.  Multiple jumps can hurt our SEO value.

Reply:
The CMS aggregates multiple jumps to 1 jump.  They are server side redirects so they would always only be one jump. It would have only a positive impact on SEO, as it would greatly reduce our 404's

We should be able to decouple title (H1) changes to the path, but that is now how things are set up now and would make management of urls confusing if they were decoupled.   This is something that Drupal has handled pretty seamlessly for a decade and a half.

Reply:
Decoupling H1s and URLs is something we have been talking about for the past year with Kevin W.  In some cases we need them decoupled to keep URLs cemented in place for SEO and communication purposes...in some cases, we may need to set up a separate field for a shortend title/description to cap characters and drive a URL...and in other cases we may have the H1s tied directly to the URLs.  Each of these is a different product and we should be able to define the right way to handle it based on business processes and access levels.